main.go

```diff
@@ -22,59 +55,232 @@ func main() {
+			}
+			selExpr, ok := callExpr.Fun.(*ast.SelectorExpr)
+			if !ok {
+				return true
+			}
+			funcObj := pkg.TypesInfo.Uses[selExpr.Sel]
+
+			signature, ok := funcObj.Type().(*types.Signature)
+			if !ok {
+				return true
+			}
+			res, ok := signature.Results().At(0).Type().(*types.Named)
+			if !ok {
+				return true
+			}
+			ures, ok := res.Underlying().(*types.Struct)
+			if !ok {
+				return true
+			}
+			futureType, ok := ures.Field(0).Type().(*types.Named)
+			if !ok {
+				return true
+			}
+			future := futureType.Obj()
+			if future.Name() != "FutureAPI" || future.Pkg().Path() != "github.com/Azure/go-autorest/autorest/azure" {
+				return true
+			}
+
+			fmt.Println("Track 1 Hit: " + strings.TrimPrefix(pkg.Fset.Position(asnmt.Pos()).String(), rootPath+string(filepath.Separator)))
+
+			return false
+		})
+	}
+}
+
+func pandoraSDKScan(pkg *packages.Package, rootPath string) {
+	funcScanCache := make(map[string][]string)
+
+	for _, f := range pkg.Syntax {
+		ast.Inspect(f, func(node ast.Node) bool {
+			asnmt, ok := node.(*ast.AssignStmt)
+			if !ok {
+				return true
+			}
+
+			if len(asnmt.Rhs) != 1 {
+				return true
+			}
+
+			callExpr, ok := asnmt.Rhs[0].(*ast.CallExpr)
+			if !ok {
+				return true
+			}
+
+			selExpr, ok := callExpr.Fun.(*ast.SelectorExpr)
+			if !ok {
+				return true
+			}
+
+			selExprName := selExpr.Sel.Name
+
+			if !strings.Contains(selExprName, "Create") &&
+				!strings.Contains(selExprName, "CreateOrUpdate") &&
+				!strings.Contains(selExprName, "Update") &&
+				!strings.Contains(selExprName, "Delete") {
+				return true
+			}
+
+			// "ThenPoll being called implies the call is always correct."
+			if strings.HasSuffix(selExprName, "ThenPoll") {
+				return true
+			}
+
+			funcObj := pkg.TypesInfo.Uses[selExpr.Sel]
+
+			_, ok = funcObj.Type().(*types.Signature)
+			if !ok {
+				return true
+			}
+
+			signature, ok := funcObj.Type().(*types.Signature)
+			if !ok {
+				return true
+			}
+
+			recv := signature.Recv()
+			if recv == nil {
+				return true
+			}
+
+			if recv.Type() == nil {
+				return true
+			}
+
+			funcRecvType := signature.Recv().Type().String()
+			funcPkg := funcObj.Pkg()
+
+			// Use cache to reduce duplicate ast traversal
+			if len(funcScanCache[funcRecvType]) > 0 {
+				pollFuncList := funcScanCache[funcRecvType]
+				for _, v := range pollFuncList {
+					if selExpr.Sel.Name+"ThenPoll" == v {
+						fmt.Println("Pandora Hit: " + strings.TrimPrefix(pkg.Fset.Position(asnmt.Pos()).String(), rootPath+string(filepath.Separator)))
+						break
+					}
 				}
-				futureType, ok := ures.Field(0).Type().(*types.Named)
-				if !ok {
-					return true
+
+				return false
+			}
+
+			// Get Pandora SDK package containing the being called synchronized CRUD function, and verify whether there is also `ThenPoll()` function.
+			// If there is, then it's bug.
+			cfgInner := &packages.Config{Mode: packages.LoadSyntax}
+			pkgsInner, err := packages.Load(cfgInner, funcPkg.Path())
```

Comment:
Can we construct this mapping at the begining of the program? That we should specify the `Mode: packages.LoadAllSyntax` to make it also load its dependencies (i.e. the go-azure-sdk package). Then you can use `.TypesInfo.Defs` in the go-azure-sdk packages to search for the `ThenPoll` functions and their sync ones, [then use the `types.Object` as the canonical values for comparing/keying](https://github.com/golang/example/blob/master/gotypes/go-types.md#:~:text=Objects%20are%20canonical,that%20uniquely%20identifies%20it.) .